chain-test-utils/src/time/mod.rs

```diff
@@ -0,0 +1,136 @@
+use chain_time::{Epoch, Slot, TimeEra};
+use smoke::{Generator, R};
+
+// Generate an `Epoch` from a generator function
+fn generate_epoch_with<GenF>(mut gen: GenF) -> Epoch
+where
+    GenF: FnMut() -> u32,
+{
+    Epoch(gen())
+}
+
+// Generate an `Slot` from a generator function
+fn generate_slot_with<GenF>(mut gen: GenF) -> Slot
+where
+    GenF: FnMut() -> u64,
+{
+    gen().into()
+}
+
+// Generate an `Epoch` given a `smoke::R` (random generator)
+pub fn generate_epoch(r: &mut R) -> Epoch {
+    generate_epoch_with(|| r.num())
+}
+
+// Generate an `Epoch` given a `smoke::R` (random generator) and a `(u32, u32)` range limit tuple
+pub fn generate_epoch_with_range(r: &mut R, range: (u32, u32)) -> Epoch {
+    generate_epoch_with(|| r.num_range(range.0, range.1))
+}
+
+// Generate an `Slot` given a `smoke::R` (random generator)
+pub fn generate_slot(r: &mut R) -> Slot {
+    generate_slot_with(|| r.num())
+}
+
+// Generate an `Slot` given a `smoke::R` (random generator) and a `(u32, u32)` range limit tuple
+pub fn generate_slot_with_range(r: &mut R, range: (u64, u64)) -> Slot {
+    generate_slot_with(|| r.num_range(range.0, range.1))
+}
+
+// Generate an `TimeEra` given a `smoke::R` (random generator)
+pub fn generate_time_era(r: &mut R) -> TimeEra {
+    TimeEra::new(generate_slot(r), generate_epoch(r), r.num())
+}
+
+// `TimeEra` configuration, encapsulates the building boundaries for the inner data
+#[derive(Clone)]
+pub struct TimeEraGenCfg {
+    pub slot_range: (u64, u64),
+    pub epoch_range: (u32, u32),
+    pub slots_per_epoch_range: (u32, u32),
+}
+
+// Generate an `TimeEra` given a `smoke::R` (random generator) and a `TimeEraGenCfg` range limit tuple
+pub fn generate_time_era_with_config(r: &mut R, config: &TimeEraGenCfg) -> TimeEra {
+    TimeEra::new(
+        generate_slot_with_range(r, config.slot_range),
+        generate_epoch_with_range(r, config.epoch_range),
+        r.num_range(
+            config.slots_per_epoch_range.0,
+            config.slots_per_epoch_range.1,
+        ),
+    )
+}
+
+// Generator wrapper for TimeEra generator methods
+pub struct TimeEraGenerator {
+    config: Option<TimeEraGenCfg>,
+}
+
+impl TimeEraGenerator {
+    pub fn new(config: Option<TimeEraGenCfg>) -> Self {
+        Self { config }
+    }
```

Comment:
what's the point to have a `new` with an optional of something and then a `with_config` to set the config?

use a `Config` and a `ConfigBuilder` if you need it. Use `Default` too. This will make your types more ergonomic.

Reply:
Well, I was thinking that we may want to change the way the generator generates values mid-way on some computation and still keep the reproducible step from the same seed. I see that `with_config` was an unfortunate name for this. It should be a `set_config`.